### PR TITLE
feature: Fractional Exponents

### PR DESCRIPTION
**Pull request by CrownedPhoenix**

### CODE REVIEW

File: Sources/Units/Unit/Fraction.swift [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+
+    init(numerator: Int, denominator: Int) {
+        let gcd = Self.gcd(numerator, denominator)
+        self.numerator = numerator / gcd
+        self.denominator = denominator / gcd
+    }
+
+    var positive: Bool {
+        switch (numerator, denominator) {
+            // 0/0 is not positive in this logic
```
Comment (NeedleInAJayStack) [inline Sources/Units/Unit/Fraction.swift]: Just as an aside, here's a fun math Wikipedia article on this particular fraction: https://en.wikipedia.org/wiki/Indeterminate_form#Indeterminate_form_0/0

File: Sources/Units/Unit/Fraction.swift [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+
+extension Fraction {
+    var asDouble: Double {
+        Double(numerator) / Double(denominator)
+    }
+}
+
+extension Fraction: Comparable {
+    public static func < (lhs: Fraction, rhs: Fraction) -> Bool {
+        return lhs.numerator < rhs.numerator
```
Comment (NeedleInAJayStack) [inline Sources/Units/Unit/Fraction.swift]: I don't think that this is correct - it should be:

```suggestion
        return lhs.asDouble < rhs.asDouble
```

File: Sources/Units/Unit/Fraction.swift [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+        if denominator == 1 {
+            "\(numerator)"
+        } else {
+            "(\(numerator)|\(denominator))"
+        }
+    }
+}
+
+extension SignedInteger {
+    func over<T: SignedInteger>(_ denominator: T) -> Fraction {
```
Comment (NeedleInAJayStack) [inline Sources/Units/Unit/Fraction.swift:169]: Were there type inference issues with using `/`, like `5.measured(in: .meter.pow(1/2))`?

File: Sources/Units/Unit/Fraction.swift [hunk 4]
```diff
@@ -0,0 +1,163 @@
+
+
+public struct Fraction: Hashable, Equatable, Sendable {
+    let numerator: Int
+    let denominator: Int
+
+    init(numerator: Int, denominator: Int) {
+        let gcd = Self.gcd(numerator, denominator)
+        self.numerator = numerator / gcd
+        self.denominator = denominator / gcd
```
Comment (NeedleInAJayStack) [inline Sources/Units/Unit/Fraction.swift:14]: Should this be public?

File: Sources/Units/Unit/Fraction.swift [hunk 5]
```diff
@@ -0,0 +1,163 @@
+
+
+public struct Fraction: Hashable, Equatable, Sendable {
```
Comment (NeedleInAJayStack) [inline Sources/Units/Unit/Fraction.swift:5]: Could we provide documentation about this type? Specific areas that would be helpful to document after reading through:

- `init` automatically reduces the input fraction
- String representation uses `|`

File: Tests/UnitsTests/UnitTests.swift [hunk 6]
```diff
@@ -7,6 +7,8 @@ final class UnitTests: XCTestCase {
         XCTAssertNotEqual(Unit.meter, Unit.foot)
         XCTAssertEqual(Unit.meter * Unit.second / Unit.second, Unit.meter)
         XCTAssertNotEqual(Unit.newton, Unit.kilogram * Unit.meter / Unit.second.pow(2))
+        XCTAssertNotEqual(Unit.meter.pow(1.over(2)), Unit.kilogram * Unit.meter / Unit.second.pow(2))
+        XCTAssertEqual(Unit.meter.pow(1.over(6)), Unit.meter.pow(1.over(2)).pow(1.over(3)) )
```
Comment (NeedleInAJayStack) [inline Tests/UnitsTests/UnitTests.swift:11]: These tests are great, but could we also add specific tests for the Fraction functions?

**Review by NeedleInAJayStack — COMMENTED** [review]
Excellent work!! I love this approach.

Just wondering - have we validated that exponents in a fractional form are sufficient? Like that defining units in a decimal-based format is not required/desired? Also worth noting (although it's probably completely irrelevant), is that integer fractions cannot represent all decimal values: https://en.wikipedia.org/wiki/Irrational_number

File: Sources/Units/Unit/Fraction.swift (hunk 2 above)
Comment (CrownedPhoenix) [reply Sources/Units/Unit/Fraction.swift]: You're right. I'll fix this. Instead of `asDouble` I'l just do the int math to create a common denominator and compare them that way.

File: Sources/Units/Unit/Fraction.swift (hunk 3 above)
Comment (CrownedPhoenix) [reply Sources/Units/Unit/Fraction.swift:169]: Yeah. That was what I originally tried.
It was ambiguous to the compiler whether `1 / 2` meant `Int / Int` or `Fraction / Int` or `Fraction / Fraction` etc because `Fraction` conforms to ExpressibleAsIntegerLiteral.

File: Sources/Units/Unit/Fraction.swift (hunk 4 above)
Comment (CrownedPhoenix) [reply Sources/Units/Unit/Fraction.swift:14]: Yes. I guess the tests do a testable import so I didn't catch that.

File: Sources/Units/Unit/Fraction.swift (hunk 5 above)
Comment (CrownedPhoenix) [reply Sources/Units/Unit/Fraction.swift:5]: Yes. Will do. I'll make a doc pass for everything I've added.

File: Tests/UnitsTests/UnitTests.swift (hunk 6 above)
Comment (CrownedPhoenix) [reply Tests/UnitsTests/UnitTests.swift:11]: Yeah. Just haven't yet cuz I didn't want to spend more time on it if it wasn't a productive direction to go. I'll make a more robust test suite for the Fraction.

**Comment by CrownedPhoenix** [comment]
> Just wondering - have we validated that exponents in a fractional form are sufficient? Like that defining units in a decimal-based format is not required/desired? Also worth noting (although it's probably completely irrelevant), is that integer fractions cannot represent all decimal values: https://en.wikipedia.org/wiki/Irrational_number

Fractional form is not sufficient. It's just a step in the right direction.
I think after this the next thing I'll work on is some sort of higher level type (maybe an enum) that subsumes both `Fraction` and `Double`. If the user enters a `Double` we assume they are both okay with double precision (as well as its limits) and that they have provided the exact value (at least to the desired significant figures) that they want used.

I haven't thought much about how we'll straddle the operation between `ft^(2/5) * s^3.5` kinds of things. I suspect there's some tricky business around that and some decisions to be made.

File: Sources/Units/Unit/Fraction.swift (hunk 3 above)
Comment (CrownedPhoenix) [reply Sources/Units/Unit/Fraction.swift:169]: On that note, how do you feel about `|` as the operator? It wouldn't be ambiguous.

You could do `2|5` etc

File: Tests/UnitsTests/UnitTests.swift (hunk 6 above)
Comment (CrownedPhoenix) [reply Tests/UnitsTests/UnitTests.swift:11]: I've added tests.

File: Sources/Units/Unit/Fraction.swift (hunk 4 above)
Comment (CrownedPhoenix) [reply Sources/Units/Unit/Fraction.swift:14]: I've made some more things public

File: Sources/Units/Unit/Fraction.swift (hunk 3 above)
Comment (CrownedPhoenix) [reply Sources/Units/Unit/Fraction.swift:169]: I went ahead and did this - lmk if you like/dislike.

File: Sources/Units/Unit/Fraction.swift (hunk 2 above)
Comment (CrownedPhoenix) [reply Sources/Units/Unit/Fraction.swift]: Updated this logic.

**Comment by NeedleInAJayStack** [comment]
> Fractional form is not sufficient. It's just a step in the right direction. I think after this the next thing I'll work on is some sort of higher level type (maybe an enum) that subsumes both `Fraction` and `Double`. If the user enters a `Double` we assume they are both okay with double precision (as well as its limits) and that they have provided the exact value (at least to the desired significant figures) that they want used.
> 
> I haven't thought much about how we'll straddle the operation between `ft^(2/5) * s^3.5` kinds of things. I suspect there's some tricky business around that and some decisions to be made.

Oof, do you think we should continue down this direction then or just try and go toward a 'double'-ized exponent? I'd hate to introduce a `Fraction` concept if we were just going to subsume its functionality with a Double soon.

**Comment by CrownedPhoenix** [comment]
> > Fractional form is not sufficient. It's just a step in the right direction. I think after this the next thing I'll work on is some sort of higher level type (maybe an enum) that subsumes both `Fraction` and `Double`. If the user enters a `Double` we assume they are both okay with double precision (as well as its limits) and that they have provided the exact value (at least to the desired significant figures) that they want used.
> > I haven't thought much about how we'll straddle the operation between `ft^(2/5) * s^3.5` kinds of things. I suspect there's some tricky business around that and some decisions to be made.
> 
> Oof, do you think we should continue down this direction then or just try and go toward a 'double'-ized exponent? I'd hate to introduce a `Fraction` concept if we were just going to subsume its functionality with a Double soon.

Sorry it won't be subsuming the functionality.
Brandon has said he needs support for both:
- `ft^(3/5)`
- `m^1.5`

The reason we need fraction parts is to preserve precision for as long as possible so we don't run into a case like `m^1.0/m^0.9999...` not cancelling right.

Can you think of any reason I wouldn't need to worry about that and could instead just convert all fractional exponents into Doubles and still get the right behavior during dimensional analysis? 

The main use case is for intermediate units that when combined reduce into units that make sense.
So an equation might be composed of units that have no real world analogue but when combined cancel out with each other appropriately to create a "real" unit.

File: Tests/UnitsTests/UnitTests.swift [hunk 7]
```diff
@@ -235,6 +276,11 @@ final class UnitTests: XCTestCase {
             Unit.meter.pow(2)
         )
 
+        XCTAssertEqual(
+            try Unit(fromSymbol: "m^(2|5)"),
+            Unit.meter.pow(2.over(5))
```
Comment (wyasul) [inline Tests/UnitsTests/UnitTests.swift:281]: Would `Unit.meter.pow(2|5)` compile with this API? As an alternative to .over

**Review by wyasul — APPROVED** [review]
Need this! Looks good.

File: Tests/UnitsTests/UnitTests.swift (hunk 7 above)
Comment (CrownedPhoenix) [reply Tests/UnitsTests/UnitTests.swift:281]: We'd have to overload the | operator which I am rather wary of doing.

Comment (CrownedPhoenix) [reply Tests/UnitsTests/UnitTests.swift:281]: Oh lol it looks like that's already what I did... haha

Comment (wyasul) [reply Tests/UnitsTests/UnitTests.swift:281]: I like that.. I think it makes intuitive sense for the digital twin developers who unfortunately write equations as strings that will need to be compiled. It's more intuitive for them to write "pow(2|5)" as a string than "pow(2.over(5))" as a string. 

In the conversation with Jay it seems like you also tried to support an API that allows pow(2/5), but this lead to some type inference issues? I think this would be ideal, but if that's not possible "|" seems appropriate

File: Sources/Units/Measurement/Measurement.swift [hunk 8]
```diff
@@ -140,9 +140,9 @@ public struct Measurement: Equatable, Codable {
     /// Exponentiate the measurement. This is equavalent to multiple `*` operations.
     /// - Parameter raiseTo: The exponent to raise the measurement to
     /// - Returns: A new measurement with an exponentiated scalar value and an exponentiated unit of measure
-    public func pow(_ raiseTo: Int) -> Measurement {
+    public func pow(_ raiseTo: Fraction) -> Measurement {
```
Comment (NeedleInAJayStack) [inline Sources/Units/Measurement/Measurement.swift:143]: To avoid a breaking change, could we just add a new function that takes `Fraction`, but keep the old one too? I think this would also resolve the current compilation failures.

File: Sources/Units/Registry.swift [hunk 9]
```diff
@@ -29,10 +29,10 @@ class Registry {
 
     /// Returns a list of defined units and their exponents, given a composite unit symbol. It is expected that the caller has
     /// verified that this is a composite unit.
-    func compositeUnitsFromSymbol(symbol: String) throws -> [DefinedUnit: Int] {
+    internal func compositeUnitsFromSymbol(symbol: String) throws -> [DefinedUnit: Fraction] {
```
Comment (NeedleInAJayStack) [inline Sources/Units/Registry.swift:32]: nit: Remove explicit `internal`

File: Sources/Units/Unit/Fraction.swift [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+
+    /// Combines the provided `numerator` and `denominator` into a reduced ``Fraction``.
+    /// - Warning: Attempts to create a ``Fraction`` with a zero denominator will fatally error.
+    public init(numerator: Int, denominator: Int) {
+        let gcd = Self.gcd(numerator, denominator)
+        self.numerator = numerator / gcd
+        self.denominator = denominator / gcd
+    }
+
+    public var positive: Bool {
```
Comment (NeedleInAJayStack) [inline Sources/Units/Unit/Fraction.swift:17]: Since we have `Comparable` conformance, do we need this? Like, couldn't we just compare an instance to zero (i.e. `fraction > 0`)? It looks like this function is only used within `description` and tests.

File: Sources/Units/Unit/Unit.swift [hunk 11]
```diff
@@ -55,7 +55,7 @@ public struct Unit {
     /// Create a new from the sub-unit dictionary.
     /// - Parameter subUnits: A dictionary of defined units and exponents. If this dictionary has only a single unit with an exponent of one,
     /// we return that defined unit directly.
-    init(composedOf subUnits: [DefinedUnit: Int]) {
+    internal init(composedOf subUnits: [DefinedUnit: Fraction]) {
```
Comment (NeedleInAJayStack) [inline Sources/Units/Unit/Unit.swift:58]: nit: Remove `internal`

File: Sources/Units/Unit/Unit.swift [hunk 12]
```diff
@@ -88,7 +88,7 @@ public struct Unit {
     public static func define(
         name: String,
         symbol: String,
-        dimension: [Quantity: Int],
+        dimension: [Quantity: Fraction],
```
Comment (NeedleInAJayStack) [inline Sources/Units/Unit/Unit.swift:91]: Same deal on all the public API changes in this file - it'd be great to make new functions instead of changing existing ones so that this isn't a breaking change. I'd expect most users to use `Int` anyway, since whole exponents are much more common than fractional.

File: Sources/Units/Unit/Fraction.swift [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+        } else {
+            return nil
+        }
+    }
+
+    public var description: String {
+        if denominator == 1 {
+            "\(!positive && numerator != 0 ? "-" : "")\(abs(numerator))"
+        } else {
+            "(\(positive ? "" : "-")\(abs(numerator))|\(abs(denominator)))"
```
Comment (NeedleInAJayStack) [inline Sources/Units/Unit/Fraction.swift:163]: I get that it isn't possible to overload the `/` Swift operator, but I think we could use it in the string representation here, right? As long as we require fractional exponents to be surrounded by parentheses, I think we can force valid parsing using `/`.

File: Sources/Units/Unit/Fraction.swift (hunk 3 above)
Comment (NeedleInAJayStack) [reply Sources/Units/Unit/Fraction.swift:169]: > Yeah. That was what I originally tried.
> It was ambiguous to the compiler whether 1 / 2 meant Int / Int or Fraction / Int or Fraction / Fraction etc because Fraction conforms to ExpressibleAsIntegerLiteral.

Ah, gotcha. Yeah, sounds good to me.

> On that note, how do you feel about | as the operator? It wouldn't be ambiguous.

For the Swift API, I don't really like it. It causes inheriting packages to require type declarations on any integer bitwise OR which can be pretty annoying (we can see this in the new FractionTests, where you had to do some gymnastics to get type declarations in there). Honestly, I'd prefer just using `over` and no operator instead of providing some syntax sugar that breaks other stuff.

For a equation/string representation, I'd like to use `/` if possible, but if not then `|` seems okay.

File: Tests/UnitsTests/FractionTests.swift [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+    func testPositive() {
+        XCTAssertTrue((3 | 5).positive)
+        XCTAssertTrue((-4 | -16).positive)
+        XCTAssertFalse((-17 | 42).positive)
+        XCTAssertFalse((1 | -111).positive)
+
+        XCTAssertTrue((0 | 1).positive)
+        XCTAssertTrue((-0 | -1).positive)
+        XCTAssertTrue((0 | -1).positive)
+        XCTAssertTrue((-0 | 1).positive)
```
Comment (NeedleInAJayStack) [inline Tests/UnitsTests/FractionTests.swift:32]: nit: Change these to ` > 0`

File: Tests/UnitsTests/UnitTests.swift [hunk 15]
```diff
@@ -116,6 +147,11 @@ final class UnitTests: XCTestCase {
             "m/s^2"
         )
 
+        XCTAssertEqual(
+            (Unit.meter / Unit.second.pow(2.over(5))).symbol,
+            "m/s^(2|5)"
+        )
```
Comment (NeedleInAJayStack) [inline Tests/UnitsTests/UnitTests.swift:153]: Could we get some fractional exponent parser tests in this MR as well? It looks like we've only added tests that verify that we can go from Swift -> String, but not from String -> Swift, and I think we haven't updated the parser to support fractional exponent equations: https://github.com/CrownedPhoenix/Units/blob/87a14f4d211172678d5f0c70c19a4a71b2d6599a/Sources/Units/Parser.swift#L166

**Review by NeedleInAJayStack — CHANGES_REQUESTED** [review]
Thanks for bringing this back up. I've added a few more requests.

File: Tests/UnitsTests/UnitTests.swift (hunk 7 above)
Comment (NeedleInAJayStack) [reply Tests/UnitsTests/UnitTests.swift:281]: I give my take here: https://github.com/NeedleInAJayStack/Units/pull/7#discussion_r2122703916

TLDR: Overloading `|` doesn't seem like a great option to me either, since it has the same `Ambiguous use` issues as `/`, except with bitwise or instead of divide.

File: Sources/Units/Unit/Equations.swift [hunk 16]
```diff
@@ -93,19 +93,19 @@ func serializeSymbolicEquation<T>(
...
     for multChunks in equation.split(separator: multSymbol, omittingEmptySubsequences: false) {
         for (index, divChunks) in multChunks.split(separator: divSymbol, omittingEmptySubsequences: false).enumerated() {
             let symbolChunks = divChunks.split(separator: expSymbol, omittingEmptySubsequences: false)
             let subSymbol = String(symbolChunks[0]).trimmingCharacters(in: .whitespaces)
-            var exp = 1
+            var exp: Fraction = 1
             if symbolChunks.count == 2 {
-                guard let expInt = Int(String(symbolChunks[1])) else {
+                guard let expInt = Fraction(String(symbolChunks[1])) else {
                     throw UnitError.invalidSymbol(message: "Symbol '^' must be followed by an integer: \(equation)")
```
Comment (NeedleInAJayStack) [inline Sources/Units/Unit/Equations.swift:109]: nit: Can we update this error message to specify that it has to be an integer or fraction?

File: Tests/UnitsTests/UnitTests.swift (hunk 15 above)
Comment (NeedleInAJayStack) [reply Tests/UnitsTests/UnitTests.swift:153]: Upon second look, we probably don't need to update the parser - that code I linked only applies to exponents of *numbers*, not *units*. 🤦

Probably still good to add some tests of parsing fractional units though.
